pkg/storage/command_queue_test.go

```diff
@@ -727,6 +728,92 @@ func TestCommandQueueTransitiveDependencies(t *testing.T) {
 	})
 }
 
+// TestCommandQueueGetSnapshotWithReadBuffer commands in the read buffer are
+// returned in the snapshot.
+func TestCommandQueueGetSnapshotWithReadBuffer(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// test that read command buffer is flushed to interval tree
+	cq := NewCommandQueue(true /* covering optimization */)
+	add(cq, roachpb.Key("a"), nil, true, nil)
+	add(cq, roachpb.Key("a"), nil, true, nil)
+
+	snapshot := cq.GetCommandQueueSnapshot()
+	if n1 := len(snapshot); n1 != 2 {
+		t.Fatalf("expected 2 commands in snapshot; got %d", n1)
+	}
+	snapshotCommands := map[int64]storagebase.CommandQueueCommand{}
+	for _, command := range snapshot {
+		snapshotCommands[command.Id] = command
+	}
+	if len(snapshotCommands[0].Prereqs) != 0 {
+		t.Fatalf("expected commands[0].Prereqs to be []; got %v", snapshotCommands[0].Prereqs)
+	}
+	if len(snapshotCommands[1].Prereqs) != 0 {
+		t.Fatalf("expected commands[1].Prereqs to be []; got %v", snapshotCommands[1].Prereqs)
+	}
+}
+
+// TestCommandQueueGetSnapshotWithChildren verifies that child commands are
+// returned in the snapshot.
+func TestCommandQueueGetSnapshotWithChildren(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	cq := NewCommandQueue(true /* covering optimization */)
+	cmd1 := add(cq, roachpb.Key("a"), nil, false, nil)
+	cmd2 := add(cq, roachpb.Key("a"), nil, true, []*cmd{cmd1})
+	// the following creates a node with two children because it has two spans
+	// only the children show up in the snapshot.
+	cq.add(true, zeroTS, []*cmd{cmd2}, []roachpb.Span{
+		{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")},
+		{Key: roachpb.Key("d"), EndKey: roachpb.Key("f")},
+	})
+
+	snapshot := cq.GetCommandQueueSnapshot()
+	if n2 := len(snapshot); n2 != 4 {
+		t.Fatalf("expected 4 commands in shapshot; got %d", n2)
+	}
+	snapshotCommands := map[int64]storagebase.CommandQueueCommand{}
+	for _, command := range snapshot {
+		snapshotCommands[command.Id] = command
+	}
+	if len(snapshotCommands[0].Prereqs) != 0 {
+		t.Fatalf("expected commands[0].Prereqs to be []; got %v", snapshotCommands[0].Prereqs)
+	}
+	if len(snapshotCommands[1].Prereqs) != 0 {
+		t.Fatalf("expected commands[1].Prereqs to be []; got %v", snapshotCommands[1].Prereqs)
+	}
+	if len(snapshotCommands[2].Prereqs) != 1 || snapshotCommands[2].Prereqs[0] != 1 {
+		t.Fatalf("expected commands[2].Prereqs to be [1]; got %v", snapshotCommands[2].Prereqs)
+	}
+	if len(snapshotCommands[3].Prereqs) != 0 {
+		t.Fatalf("expected commands[3].Prereqs to be []; got %v", snapshotCommands[3].Prereqs)
```

Comment:
@nvanbenschoten I'm surprised by this output… Shouldn't this command have a prereq on the command with id 1? (It's one of the leaf commands generated when a command with two spans is added)

Reply:
Planning to give all these changes a more thorough review tomorrow. For now, double check what `snapshotCommands[3]` is giving you. `snapshotCommands` is indexed by the command id, not the index of the command in the snapshot slice. I don't think there is a command with id 0 or 3 so I think you're seeing default values for some of these. I'd expect ids `[1, 2, 4, 5]` (gap because of covering span).

Also, these tests should assert as much as possible. Try something like:
```
if expPrereqs := []int64{1}; !reflect.DeepEqual(expPrereqs, snapshotCommands[5].Prereqs) {
    ...
}
```